bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/ICalendarHandlerFactory.java

```diff
@@ -67,6 +84,61 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
         if (!supportsThingType(thingTypeUID)) {
             return null;
         }
-        return new ICalendarHandler(thing, sharedHttpClient, eventPublisher);
+        if (thingTypeUID.equals(THING_TYPE_CALENDAR)) {
+            if (thing instanceof Bridge) {
+                return new ICalendarHandler((Bridge) thing, sharedHttpClient, eventPublisher);
+            } else {
+                // Migration needs to be done asynchronously. Using thread pool for common things.
+                ThingUID uidToCopy = thing.getUID();
+                ExecutorService threadPool = ThreadPoolManager.getPool(ThreadPoolManager.THREAD_POOL_NAME_COMMON);
+                threadPool.execute(new ReregisterThingRunnable(uidToCopy));
+                return null;
+            }
+        } else if (thingTypeUID.equals(THING_TYPE_FILTERED_EVENTS)) {
+            return new EventFilterHandler(thing);
+        }
+        return null;
+    }
+
+    /**
+     * This Runnable "upgrades" the thing "calendar" to the bridge of same type. To make sure the thing isn't in use, a
+     * sleep blocks before doing actual work.
+     *
+     * @author Michael Wodniok - Initial contribution
+     */
+    private class ReregisterThingRunnable implements Runnable {
+        private final ThingUID thingUID;
+
+        public ReregisterThingRunnable(ThingUID uid) {
+            thingUID = uid;
+        }
+
+        @Override
+        public void run() {
+            logger.info("Converting Thing {} to Bridge. This will remove the item and readd it. Please stand by.",
+                    thingUID);
+            try {
+                Thread.sleep(20000);
```

Comment:
Why is this delay necessary?

Reply:
OpenHAB does not provide hooks for upgrades AFAIK. It is as i need to be sure that the replacement is done asynchronous after the failed initialization. So `createHandler(Thing)` must return before i'm able to do the needed migration - which does not allow to use usual synchronization mechanisms. I'm pretty sure a 20 sec delay is enough to let the method return (it is just the next instruction after starting the thread containing the sleep).

Reply:
What happens if you do it synchronously? Genuinely asking.

Reply:
I've tried that first a long time ago. I couldn't add an item with same id as it collides with the existing one (that happens while migration). I think the removal has been not possible in the state "initializing" and the framework does not seem to remove the item in that time (all AFIAR!).

Reply:
@openhab/add-ons-maintainers is there a pattern to migrate Things from one version to another? In this case a Thread is spawn in the ThingHandlerFactory and after some delay the Thing is removed and re-added. Is this a valid approach?

Reply:
No, there is not. But this should definitely be done by the framework. We are looking for a solution for a long time (see https://github.com/eclipse-archived/smarthome/issues/2555). There are already drafts for an implementation like https://github.com/openhab/openhab-core/pull/1027 but unfortunately not THE result we are looking for.

Reply:
@cweitkamp The question is now: how to continue here?

Reply:
I'd clearly say: Remove any migration logic from the binding (I never accepted that in other bindings either).
As long as there is no solution in the core framework, add a note to the distro that there is a breaking change in the binding configuration and that Things have to be re-created by the user.

Reply:
hmm... ok.

Reply:
I understand that this is not what you wanted to hear, sorry for that - but I always prefer general solutions to problems (even if they are in the future) instead of having 200 bindings trying to solve it on their own in different ways. So thanks for your understanding!